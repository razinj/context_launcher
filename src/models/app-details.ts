export type AppDetails = {
  label: string
  name: string
}

export type AppDetailsWithOptionalIcon = AppDetails & { icon?: string }

export type AppDetailsWithIcon = AppDetails & { icon: string }
