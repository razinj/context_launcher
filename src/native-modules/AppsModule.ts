import { NativeModules } from 'react-native'

export interface AppsModuleInterface {
  getConstants(): AppsModuleConstants
  launchApplication(packageName: string): void
  showApplicationDetails(packageName: string): void
  requestApplicationUninstall(packageName: string): void
  getApplications(): Promise<string>
}

type AppsModuleConstants = {
  appVersion: string
  buildNumber: string
  packageName: string
}

const { AppsModule } = NativeModules

export default AppsModule as AppsModuleInterface
