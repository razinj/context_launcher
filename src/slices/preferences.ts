// Redux
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
// Models
import { RootState } from '../store'

export interface PreferencesState {
  displayRecentApps: boolean
  displayFavoriteApps: boolean
}

const initialState: PreferencesState = {
  displayRecentApps: true,
  displayFavoriteApps: true,
}

export const preferencesSlice = createSlice({
  name: 'preferences',
  initialState,
  reducers: {
    displayRecentApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayRecentApps = payload
    },
    displayFavoriteApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayFavoriteApps = payload
    },
    resetPreferences: (state: PreferencesState) => {
      state.displayRecentApps = true
      state.displayFavoriteApps = true
    },
  },
})

export const { displayRecentApps, displayFavoriteApps, resetPreferences } = preferencesSlice.actions

const selectDisplayRecentApps = (state: RootState) => state.preferences.displayRecentApps
const selectDisplayFavoriteApps = (state: RootState) => state.preferences.displayFavoriteApps

export const selectDisplayRecentAppsMemoized = createSelector(selectDisplayRecentApps, (value: boolean) => value)
export const selectDisplayFavoriteAppsMemoized = createSelector(selectDisplayFavoriteApps, (value: boolean) => value)

export default preferencesSlice.reducer
