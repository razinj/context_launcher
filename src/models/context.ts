// React
import { RefObject } from 'react'
// React Native
import { TextInput } from 'react-native'
// BottomSheet
import { BottomSheetModal } from '@gorhom/bottom-sheet'
// Models
import { AppDetailsWithOptionalIcon } from '../models/app-details'

export type GlobalContextType = {
  // All apps
  hideAllApps: () => void
  displayAllApps: boolean
  toggleDisplayAllApps: () => void
  // Settings bottom sheet
  settingsBottomSheetRef: RefObject<BottomSheetModal> | null
  displaySettingsBottomSheet: () => void
  // App item menu
  appItemMenuDetails: AppDetailsWithOptionalIcon | null
  setAppItemMenuDetails: (appDetails: AppDetailsWithOptionalIcon) => void
  appItemMenuBottomSheetRef: RefObject<BottomSheetModal> | null
  displayAppItemMenuBottomSheet: () => void
  // Favorite apps sort
  sortableFavoriteApps: boolean
  toggleSortableFavoriteApps: () => void
  // Global app launch procedure
  globalAppLaunchProcedure: () => void
  // Misc
  dismissKeyboard: () => void
}

export type SearchContextType = {
  // Search input
  searchInputRef: RefObject<TextInput> | null
  searchAppLaunchProcedure: () => void
  // Search query validity
  invalidCharacters: boolean
  setInvalidCharacters: (isInvalid: boolean) => void
}
