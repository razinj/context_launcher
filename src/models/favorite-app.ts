import { AppDetails } from './app-details'

export type FavoriteApp = AppDetails
