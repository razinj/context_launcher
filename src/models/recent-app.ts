import { AppDetails } from './app-details'

export type RecentApp = AppDetails // TODO: Remove this type?
