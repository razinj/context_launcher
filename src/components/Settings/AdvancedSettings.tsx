// React
import React from 'react'
// React Native
import { Pressable, PressableAndroidRippleConfig, StyleSheet, ToastAndroid, View } from 'react-native'
// Components
import SettingsItemLabel from './SettingsItemLabel'
// Redux
import { useDispatch } from 'react-redux'
import { setAppsList } from '../../slices/appsList'
import { clearRecentApps } from '../../slices/recentApps'
import { resetPreferences } from '../../slices/preferences'
import { clearFavoriteApps } from '../../slices/favoriteApps'
// Native modules
import AppsModule from '../../native-modules/AppsModule'
// Models
import { AppDetails } from '../../models/app-details'

const settingItemButtonRippleConfig: PressableAndroidRippleConfig = {
  borderless: false,
  foreground: true,
  color: '#e5e5e5',
}

const AdvancedSettings = () => {
  const dispatch = useDispatch()

  const reloadAllApps = () => {
    AppsModule.getApplications((applications: string) => {
      const apps = JSON.parse(applications) as AppDetails[]

      dispatch(setAppsList(apps))
      displayToast('All apps reloaded successfully!')
    })
  }

  const onClearRecentApps = () => {
    dispatch(clearRecentApps())
    displayToast('Recent apps cleared successfully!')
  }

  const onClearFavoriteApps = () => {
    dispatch(clearFavoriteApps())
    displayToast('Favorite apps cleared successfully!')
  }

  const onResetPreferences = () => {
    dispatch(resetPreferences())
    displayToast('Settings reset successfully!')
  }

  const displayToast = (message: string, duration: number = ToastAndroid.LONG) => {
    ToastAndroid.show(message, duration)
  }

  return (
    <>
      {/* Reload all apps */}
      <View style={styles.itemContainer}>
        <Pressable
          testID='advanced-settings-reload-all-apps-button'
          onPress={reloadAllApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.buttonItemPressable}>
          <SettingsItemLabel title='Reload all apps' />
        </Pressable>
      </View>

      {/* Clear recent apps */}
      <View style={styles.itemContainer}>
        <Pressable
          testID='advanced-settings-clear-recent-apps-button'
          onPress={onClearRecentApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.buttonItemPressable}>
          <SettingsItemLabel title='Clear recent apps' />
        </Pressable>
      </View>

      {/* Clear favorite apps */}
      <View style={styles.itemContainer}>
        <Pressable
          testID='advanced-settings-clear-favorite-apps-button'
          onPress={onClearFavoriteApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.buttonItemPressable}>
          <SettingsItemLabel title='Clear favorite apps' />
        </Pressable>
      </View>

      {/* Reset preferences */}
      <View style={styles.itemContainer}>
        <Pressable
          testID='advanced-settings-reset-preferences-button'
          onPress={onResetPreferences}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.buttonItemPressable}>
          <SettingsItemLabel title='Reset settings' />
        </Pressable>
      </View>
    </>
  )
}

const styles = StyleSheet.create({
  itemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    borderBottomWidth: 1,
    borderBottomColor: '#e5e5e5',
  },
  buttonItemPressable: {
    flex: 1,
  },
})

export default AdvancedSettings
