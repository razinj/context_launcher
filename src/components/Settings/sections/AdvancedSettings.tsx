import React from 'react'
import { View } from 'react-native'
import { useDispatch } from 'react-redux'
import { getAppsListAction } from '../../../slices/appsList'
import { resetPreferences } from '../../../slices/preferences'
import { displayToast } from '../../../utils/toast'
import CustomPressable from '../../shared/CustomPressable'
import SettingsItemLabel from '../shared/SettingsItemLabel'
import { settingItemButtonRippleConfig, settingItemWrapperStyle, settingsPressableItemStyle } from '../shared/values'

const AdvancedSettings = () => {
  const dispatch = useDispatch()

  const reloadAllApps = () => {
    dispatch(getAppsListAction())
    displayToast('All apps reloaded successfully!')
  }

  const onResetPreferences = () => {
    dispatch(resetPreferences())
    displayToast('Settings reset successfully!')
  }

  return (
    <>
      <View style={settingItemWrapperStyle}>
        <CustomPressable
          testID='reload-all-apps-button'
          onPress={reloadAllApps}
          android_ripple={settingItemButtonRippleConfig}
          style={settingsPressableItemStyle}>
          <SettingsItemLabel title='Reload all apps' />
        </CustomPressable>
      </View>

      <View style={settingItemWrapperStyle}>
        <CustomPressable
          testID='reset-preferences-button'
          onPress={onResetPreferences}
          android_ripple={settingItemButtonRippleConfig}
          style={settingsPressableItemStyle}>
          <SettingsItemLabel title='Reset settings' />
        </CustomPressable>
      </View>
    </>
  )
}

export default AdvancedSettings
