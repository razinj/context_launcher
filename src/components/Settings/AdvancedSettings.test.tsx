import { fireEvent, screen } from '@testing-library/react-native'
import { initialStoreState } from '../../../utils/test/data'
import { renderWithProvider } from '../../../utils/test/utils'
import { FavoriteApp } from '../../models/favorite-app'
import { RecentAppDetails } from '../../models/recent-app'
import { PreferencesState } from '../../slices/preferences'
import { ToastAndroid } from 'react-native'
import AppsModule from '../../native-modules/AppsModule'
import AdvancedSettings from './AdvancedSettings'

describe('<AdvancedSettings /> Tests', () => {
  beforeEach(() => {
    jest.spyOn(ToastAndroid, 'show')
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should render correctly and match snapshot', () => {
    renderWithProvider(<AdvancedSettings />)

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('advanced-settings-reload-all-apps-button')).toBeDefined()
    expect(screen.getByTestId('advanced-settings-clear-recent-apps-button')).toBeDefined()
    expect(screen.getByTestId('advanced-settings-clear-favorite-apps-button')).toBeDefined()
    expect(screen.getByTestId('advanced-settings-reset-preferences-button')).toBeDefined()
  })

  // TODO: Find a way to mock internal dispatched action
  it.todo('should call AppsModule method to get all applications and set apps list')

  it('should call AppsModule method to get all applications', () => {
    renderWithProvider(<AdvancedSettings />)

    const reloadAllAppsButton = screen.getByTestId('advanced-settings-reload-all-apps-button')

    expect(reloadAllAppsButton).toBeDefined()

    fireEvent.press(reloadAllAppsButton)

    expect(AppsModule.getApplications).toBeCalled()
  })

  it('should clear recent apps when button is pressed', () => {
    const customInitialState = {
      ...initialStoreState,
      recentApps: {
        list: [
          {
            name: 'com.google.chrome',
            label: 'Chrome',
            icon: 'icon-base64-string-chrome',
          },
          {
            name: 'com.google.maps',
            label: 'Maps',
            icon: 'icon-base64-string-maps',
          },
        ] as RecentAppDetails[],
      },
    }

    const { store } = renderWithProvider(<AdvancedSettings />, { preloadedState: customInitialState })

    const clearRecentAppsButton = screen.getByTestId('advanced-settings-clear-recent-apps-button')

    expect(clearRecentAppsButton).toBeDefined()

    fireEvent.press(clearRecentAppsButton)

    expect(store.getState().recentApps.list).toEqual([])
    expect(ToastAndroid.show).toBeCalledWith('Recent apps cleared successfully!', ToastAndroid.LONG)
  })

  it('should clear favorite apps when button is pressed', () => {
    const customInitialState = {
      ...initialStoreState,
      favoriteApps: {
        list: [
          {
            name: 'com.google.chrome',
            label: 'Chrome',
            icon: 'icon-base64-string-chrome',
          },
          {
            name: 'com.google.maps',
            label: 'Maps',
            icon: 'icon-base64-string-maps',
          },
        ] as FavoriteApp[],
      },
    }

    const { store } = renderWithProvider(<AdvancedSettings />, { preloadedState: customInitialState })

    const clearFavoriteAppsButton = screen.getByTestId('advanced-settings-clear-favorite-apps-button')

    expect(clearFavoriteAppsButton).toBeDefined()

    fireEvent.press(clearFavoriteAppsButton)

    expect(store.getState().favoriteApps.list).toEqual([])
    expect(ToastAndroid.show).toBeCalledWith('Favorite apps cleared successfully!', ToastAndroid.LONG)
  })

  it('should reset preferences apps when button is pressed', () => {
    const customInitialState = {
      ...initialStoreState,
      preferences: {
        displayRecentApps: false,
        displayFavoriteApps: true,
      } as PreferencesState,
    }

    const { store } = renderWithProvider(<AdvancedSettings />, { preloadedState: customInitialState })

    const resetPreferencesButton = screen.getByTestId('advanced-settings-reset-preferences-button')

    expect(resetPreferencesButton).toBeDefined()

    fireEvent.press(resetPreferencesButton)

    expect(store.getState().preferences).toEqual({
      displayRecentApps: true,
      displayFavoriteApps: true,
    })
    expect(ToastAndroid.show).toBeCalledWith('Settings reset successfully!', ToastAndroid.LONG)
  })
})
