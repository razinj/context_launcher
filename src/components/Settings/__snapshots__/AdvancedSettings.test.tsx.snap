// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdvancedSettings /> Tests should render correctly and match snapshot 1`] = `
[
  <View
    style={
      {
        "borderBottomColor": "#e5e5e5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
        }
      }
      testID="advanced-settings-reload-all-apps-button"
    >
      <View
        style={
          {
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#808080",
            }
          }
        >
          Reload all apps
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "borderBottomColor": "#e5e5e5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
        }
      }
      testID="advanced-settings-clear-recent-apps-button"
    >
      <View
        style={
          {
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#808080",
            }
          }
        >
          Clear recent apps
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "borderBottomColor": "#e5e5e5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
        }
      }
      testID="advanced-settings-clear-favorite-apps-button"
    >
      <View
        style={
          {
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#808080",
            }
          }
        >
          Clear favorite apps
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "borderBottomColor": "#e5e5e5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
        }
      }
      testID="advanced-settings-reset-preferences-button"
    >
      <View
        style={
          {
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#808080",
            }
          }
        >
          Reset settings
        </Text>
      </View>
    </View>
  </View>,
]
`;
